cache/in-memory/src/lib.rs

```diff
@@ -111,31 +115,89 @@ use twilight_model::{
     voice::VoiceState,
 };
 
-#[derive(Debug)]
-struct GuildItem<T> {
-    data: T,
+/// Resource associated with a guild.
+///
+/// This is used when a resource does not itself include its associated guild's
+/// ID. In lieu of the resource itself storing its guild's ID this relation
+/// includes it.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct GuildResource<T> {
     guild_id: GuildId,
+    value: T,
+}
+
+impl<T> GuildResource<T> {
+    /// ID of the guild associated with the resource.
+    pub const fn guild_id(&self) -> GuildId {
+        self.guild_id
+    }
+
+    /// Immutable reference to the resource's value.
+    pub const fn resource(&self) -> &T {
+        &self.value
+    }
+}
+
+impl<T> Deref for GuildResource<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        self.resource()
+    }
+}
+
+/// Immutable reference to a resource in the cache.
+// We need this so as not to expose the underlying cache implementation.
+pub struct Reference<'a, K, V> {
```

Comment:
In theory, this could implement `From<Ref>` and be used in all the places it returns in place of maps, correct?

Reply:
I'm not sure what you mean by "in place of maps," but it can't implement `From` for `Ref` because that would expose `dashmap` as a dependency.

Reply:
I didn't phrase that correctly, but you're reasoning makes sense